src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala

```diff
@@ -0,0 +1,477 @@
+          .ensure(s"Fetched node state hash doesn't match requested one, blacklisting peer")(nodes =>
+            fetcher.hash == kec256(nodes.head))
+
+        validatedNode match {
+          case Left(err) =>
+            log.debug(err)
+            peersClient ! BlacklistPeer(peer.id, err)
+            fetchStateNode(fetcher.hash, fetcher.replyTo, state)
+          case Right(node) =>
+            fetcher.replyTo ! FetchedStateNode(NodeData(node))
+            context become started(state.notFetchingStateNode())
+        }
+      })
+  }
+
+  private def handleNewBlockMessages(state: FetcherState): Receive = {
+    case MessageFromPeer(NewBlockHashes(hashes), _) =>
+      val newState = state.validatedHashes(hashes) match {
+        case Left(_) => state
+        case Right(validHashes) => state.withPossibleNewTopAt(validHashes.lastOption.map(_.number))
+      }
+
+      fetchBlocks(newState)
+    case MessageFromPeer(NewBlock(block, _), peerId) =>
+      val newBlockNr = Block.number(block)
+      val nextExpectedBlock = state.lastFullBlockNumber + 1
+
+      log.debug("Received NewBlock nr {}", newBlockNr)
+
+      // we're on top, so we can pass block directly to importer
+      if (newBlockNr == nextExpectedBlock && state.isOnTop) {
+        val newState = state.withPeerForBlocks(peerId, Seq(newBlockNr)).withKnownTopAt(newBlockNr)
+        state.importer ! OnTop
+        state.importer ! ImportNewBlock(block, peerId)
+        context become started(newState)
+        // there are some blocks waiting for import but it seems that we reached top on fetch side so we can enqueue new block for import
+      } else if (newBlockNr == nextExpectedBlock && !state.isFetching && state.waitingHeaders.isEmpty) {
+        val newState = state.appendNewBlock(block, peerId)
+        context become started(newState)
+        // waiting for some bodies but we don't have this header yet - at least we can use new block header
+      } else if (newBlockNr == state.lastBlock + 1 && !state.isFetchingHeaders) {
+        state.appendHeaders(List(block.header)) |> fetchBlocks
+        // we're far from top
+      } else if (newBlockNr > nextExpectedBlock) {
+        context become started(state.withKnownTopAt(newBlockNr))
+      }
+    case BlockImportFailed(blockNr, reason) =>
+      val (peerId, newState) = state.invalidateBlocksFrom(blockNr)
+      peerId.foreach(id => peersClient ! BlacklistPeer(id, reason))
+      fetchBlocks(newState)
+  }
+
+  private def handlePickedBlocks(state: FetcherState)(
+      pickResult: Option[(NonEmptyList[Block], FetcherState)]): FetcherState =
+    pickResult
+      .tap({
+        case (blocks, newState) =>
+          sender() ! PickedBlocks(blocks)
+          newState.importer ! (if (newState.isOnTop) OnTop else NotOnTop)
+      })
+      .fold(state)(_._2)
+
+  private def fetchBlocks(state: FetcherState): Unit = {
+    log.debug("Fetching blocks")
+    val newState = state |> tryFetchHeaders |> tryFetchBodies
```

Comment:
I have mixed feeling about this construct, it looks like simple function application but actually it makes requests, schedule retries etc.  May export it to separate function and add some scala doc explain what is happening behind scenes ?